### ISSUE TITLE
Crash in raft branch during shutdown in shard_fuzzer.py (Guarantee failed: [before != __null])

### ISSUE THREAD

**Issue by Tryneus** [rethinkdb/rethinkdb#4262]
```
error: Error in ./src/containers/intrusive_list.hpp at line 175:
error: Guarantee failed: [before != __null]
error: Backtrace:
error: Thu May 21 18:25:23 2015
  1: rethinkdb_backtrace(void**, int) at rethinkdb_backtrace.cc:100
  2: backtrace_t::backtrace_t() at backtrace.cc:203
  3: lazy_backtrace_formatter_t::lazy_backtrace_formatter_t() at backtrace.cc:283
  4: format_backtrace(bool) at backtrace.cc:198
  5: report_fatal_error(char const*, int, char const*, ...) at errors.cc:83
  6: intrusive_list_t<publisher_t<signal_t::subscription_t*>::subscription_t>::insert_between(publisher_t<signal_t::subscription_t*>::subscription_t*, intrusive_list_node_t<publisher_t<signal_t::subscription_t*>::subscription_t>*, intrusive_list_node_t<publisher_t<signal_t::subscription_t*>::subscription_t>*) at intrusive_list.hpp:175
  7: intrusive_list_t<publisher_t<signal_t::subscription_t*>::subscription_t>::push_back(publisher_t<signal_t::subscription_t*>::subscription_t*) at intrusive_list.hpp:111
  8: publisher_t<signal_t::subscription_t*>::subscription_t::reset(publisher_t<signal_t::subscription_t*>*) at pubsub.hpp:58
  9: signal_t::subscription_t::reset(signal_t*) at signal.hpp:61
  10: wait_any_t::add(signal_t const*) at wait_any.cc:30
  11: wait_any_t::wait_any_t<signal_t const*, signal_t const*>(signal_t const*, signal_t const*) at wait_any.hpp:18
  12: wait_any_t::wait_any_t<signal_t const*, signal_t const*>(signal_t const*, signal_t const*) at wait_any.hpp:19
  13: wait_interruptible(signal_t const*, signal_t const*) at interruptor.cc:7
  14: /home/ssd2/grey/rethinkdb/build/debug_clang/rethinkdb() [0x1e3f2f7] at 0x1e3f2f7 ()
  15: /home/ssd2/grey/rethinkdb/build/debug_clang/rethinkdb() [0x1e3f14c] at 0x1e3f14c ()
  16: callable_action_wrapper_t::run() at runtime_utils.cc:43
  17: coro_t::run() at coroutines.cc:205
  18: /home/ssd2/grey/rethinkdb/build/debug_clang/rethinkdb() [0x1e3ef1d] at 0x1e3ef1d ()
  19: multi_table_manager_t::schedule_sync(uuid_u const&, multi_table_manager_t::table_t*, peer_id_t const&) at multi_table_manager.cc:589
  20: /home/ssd2/grey/rethinkdb/build/debug_clang/rethinkdb() [0x1e41427] at 0x1e41427 ()
  21: /home/ssd2/grey/rethinkdb/build/debug_clang/rethinkdb() [0x1e410f7] at 0x1e410f7 ()
  22: std::function<void (peer_id_t const&, multi_table_manager_bcard_t const*)>::operator()(peer_id_t const&, multi_table_manager_bcard_t const*) const at functional:2160
  23: watchable_map_transform_t<peer_id_t, cluster_directory_metadata_t, peer_id_t, multi_table_manager_bcard_t>::read_all(std::function<void (peer_id_t const&, multi_table_manager_bcard_t const*)> const&)::{lambda(peer_id_t const&, cluster_directory_metadata_t const*)#1}::operator()(peer_id_t const&, cluster_directory_metadata_t const*) const at watchable_transform.tcc:70
  24: std::_Function_handler<void (peer_id_t const&, cluster_directory_metadata_t const*), watchable_map_transform_t<peer_id_t, cluster_directory_metadata_t, peer_id_t, multi_table_manager_bcard_t>::read_all(std::function<void (peer_id_t const&, multi_table_manager_bcard_t const*)> const&)::{lambda(peer_id_t const&, cluster_directory_metadata_t const*)#1}>::_M_invoke(std::_Any_data const&, peer_id_t const&, cluster_directory_metadata_t const*) at functional:1779
  25: std::function<void (peer_id_t const&, cluster_directory_metadata_t const*)>::operator()(peer_id_t const&, cluster_directory_metadata_t const*) const at functional:2160
  26: watchable_map_var_t<peer_id_t, cluster_directory_metadata_t>::read_all(std::function<void (peer_id_t const&, cluster_directory_metadata_t const*)> const&) at watchable_map.tcc:202
  27: watchable_map_transform_t<peer_id_t, cluster_directory_metadata_t, peer_id_t, multi_table_manager_bcard_t>::read_all(std::function<void (peer_id_t const&, multi_table_manager_bcard_t const*)> const&) at watchable_transform.tcc:62
  28: multi_table_manager_t::active_table_t::on_raft_commit() at multi_table_manager.cc:199
  29: std::_Mem_fn<void (multi_table_manager_t::active_table_t::*)()>::operator()(multi_table_manager_t::active_table_t*) const at functional:550
  30: void std::_Bind<std::_Mem_fn<void (multi_table_manager_t::active_table_t::*)()> (multi_table_manager_t::active_table_t*)>::__call<void, , 0>(std::tuple<>&&, std::_Index_tuple<0>) at functional:1145
  31: void std::_Bind<std::_Mem_fn<void (multi_table_manager_t::active_table_t::*)()> (multi_table_manager_t::active_table_t*)>::operator()<, void>() at functional:1203
  32: std::_Function_handler<void (), std::_Bind<std::_Mem_fn<void (multi_table_manager_t::active_table_t::*)()> (multi_table_manager_t::active_table_t*)> >::_M_invoke(std::_Any_data const&) at functional:1779
  33: std::function<void ()>::operator()() const at functional:2160
  34: call_function(std::function<void ()> const&) at watchable.hpp:185
  35: void publisher_controller_t<std::function<void ()> >::publish<void (*)(std::function<void ()> const&)>(void (* const&)(std::function<void ()> const&)) at pubsub.hpp:139
  36: watchable_variable_t<raft_member_t<table_raft_state_t>::state_and_config_t>::apply_atomic_op(std::function<bool (raft_member_t<table_raft_state_t>::state_and_config_t*)> const&) at watchable.hpp:223
  37: raft_member_t<table_raft_state_t>::update_commit_index(unsigned long, new_mutex_acq_t const*) at raft_core.tcc:1061
  38: raft_member_t<table_raft_state_t>::leader_update_match_index(raft_member_id_t, unsigned long, new_mutex_acq_t const*, signal_t*) at raft_core.tcc:1142
  39: raft_member_t<table_raft_state_t>::leader_send_updates(raft_member_id_t const&, unsigned long, auto_drainer_t::lock_t) at raft_core.tcc:1794
  40: std::_Mem_fn<void (raft_member_t<table_raft_state_t>::*)(raft_member_id_t const&, unsigned long, auto_drainer_t::lock_t)>::operator()(raft_member_t<table_raft_state_t>*, raft_member_id_t const&, unsigned long, auto_drainer_t::lock_t) const at functional:550
  41: void std::_Bind<std::_Mem_fn<void (raft_member_t<table_raft_state_t>::*)(raft_member_id_t const&, unsigned long, auto_drainer_t::lock_t)> (raft_member_t<table_raft_state_t>*, raft_member_id_t, unsigned long, auto_drainer_t::lock_t)>::__call<void, , 0, 1, 2, 3>(std::tuple<>&&, std::_Index_tuple<0, 1, 2, 3>) at functional:1145
  42: void std::_Bind<std::_Mem_fn<void (raft_member_t<table_raft_state_t>::*)(raft_member_id_t const&, unsigned long, auto_drainer_t::lock_t)> (raft_member_t<table_raft_state_t>*, raft_member_id_t, unsigned long, auto_drainer_t::lock_t)>::operator()<, void>() at functional:1203
  43: callable_action_instance_t<std::_Bind<std::_Mem_fn<void (raft_member_t<table_raft_state_t>::*)(raft_member_id_t const&, unsigned long, auto_drainer_t::lock_t)> (raft_member_t<table_raft_state_t>*, raft_member_id_t, unsigned long, auto_drainer_t::lock_t)> >::run_action() at callable_action.hpp:31
  44: callable_action_wrapper_t::run() at runtime_utils.cc:43
  45: coro_t::run() at coroutines.cc:205
  46: coro_t* coro_t::spawn_sometime<std::_Bind<std::_Mem_fn<void (raft_member_t<table_raft_state_t>::*)(raft_member_id_t const&, unsigned long, auto_drainer_t::lock_t)> (raft_member_t<table_raft_state_t>*, raft_member_id_t, unsigned long, auto_drainer_t::lock_t)> >(std::_Bind<std::_Mem_fn<void (raft_member_t<table_raft_state_t>::*)(raft_member_id_t const&, unsigned long, auto_drainer_t::lock_t)> (raft_member_t<table_raft_state_t>*, raft_member_id_t, unsigned long, auto_drainer_t::lock_t)>&&) at coroutines.hpp:58
  47: raft_member_t<table_raft_state_t>::leader_spawn_update_coros(unsigned long, std::map<raft_member_id_t, scoped_ptr_t<auto_drainer_t>, std::less<raft_member_id_t>, std::allocator<std::pair<raft_member_id_t const, scoped_ptr_t<auto_drainer_t> > > >*, new_mutex_acq_t const*) at raft_core.tcc:1575
  48: raft_member_t<table_raft_state_t>::candidate_and_leader_coro(new_mutex_acq_t*, auto_drainer_t::lock_t) at raft_core.tcc:1325
  49: std::_Mem_fn<void (raft_member_t<table_raft_state_t>::*)(new_mutex_acq_t*, auto_drainer_t::lock_t)>::operator()(raft_member_t<table_raft_state_t>*, new_mutex_acq_t*, auto_drainer_t::lock_t) const at functional:550
  50: void std::_Bind<std::_Mem_fn<void (raft_member_t<table_raft_state_t>::*)(new_mutex_acq_t*, auto_drainer_t::lock_t)> (raft_member_t<table_raft_state_t>*, new_mutex_acq_t*, auto_drainer_t::lock_t)>::__call<void, , 0, 1, 2>(std::tuple<>&&, std::_Index_tuple<0, 1, 2>) at functional:1145
  51: void std::_Bind<std::_Mem_fn<void (raft_member_t<table_raft_state_t>::*)(new_mutex_acq_t*, auto_drainer_t::lock_t)> (raft_member_t<table_raft_state_t>*, new_mutex_acq_t*, auto_drainer_t::lock_t)>::operator()<, void>() at functional:1203
  52: callable_action_instance_t<std::_Bind<std::_Mem_fn<void (raft_member_t<table_raft_state_t>::*)(new_mutex_acq_t*, auto_drainer_t::lock_t)> (raft_member_t<table_raft_state_t>*, new_mutex_acq_t*, auto_drainer_t::lock_t)> >::run_action() at callable_action.hpp:31
  53: callable_action_wrapper_t::run() at runtime_utils.cc:43
  54: coro_t::run() at coroutines.cc:205
  55: coro_t* coro_t::spawn_sometime<std::_Bind<std::_Mem_fn<void (raft_member_t<table_raft_state_t>::*)(new_mutex_acq_t*, auto_drainer_t::lock_t)> (raft_member_t<table_raft_state_t>*, new_mutex_acq_t*, auto_drainer_t::lock_t)> >(std::_Bind<std::_Mem_fn<void (raft_member_t<table_raft_state_t>::*)(new_mutex_acq_t*, auto_drainer_t::lock_t)> (raft_member_t<table_raft_state_t>*, new_mutex_acq_t*, auto_drainer_t::lock_t)>&&) at coroutines.hpp:58
  56: raft_member_t<table_raft_state_t>::on_watchdog_timer()::{lambda()#1}::operator()() const at raft_core.tcc:991
  57: callable_action_instance_t<raft_member_t<table_raft_state_t>::on_watchdog_timer()::{lambda()#1}>::run_action() at callable_action.hpp:31
  58: callable_action_wrapper_t::run() at runtime_utils.cc:43
  59: coro_t::run() at coroutines.cc:205
  60: coro_t* coro_t::spawn_sometime<raft_member_t<table_raft_state_t>::on_watchdog_timer()::{lambda()#1}>(raft_member_t<table_raft_state_t>::on_watchdog_timer()::{lambda()#1}&&) at coroutines.hpp:58
  61: raft_member_t<table_raft_state_t>::on_watchdog_timer() at raft_core.tcc:969
  62: raft_member_t<table_raft_state_t>::raft_member_t(raft_member_id_t const&, raft_storage_interface_t<table_raft_state_t>*, raft_network_interface_t<table_raft_state_t>*, raft_persistent_state_t<table_raft_state_t> const&, std::string const&)::{lambda()#1}::operator()() const at raft_core.tcc:92
  63: std::_Function_handler<void (), raft_member_t<table_raft_state_t>::raft_member_t(raft_member_id_t const&, raft_storage_interface_t<table_raft_state_t>*, raft_network_interface_t<table_raft_state_t>*, raft_persistent_state_t<table_raft_state_t> const&, std::string const&)::{lambda()#1}>::_M_invoke(std::_Any_data const&) at functional:1779
  64: std::function<void ()>::operator()() const at functional:2160
  65: call_ringer(std::function<void ()>) at timing.cc:93
  66: void std::_Bind<void (*(std::function<void ()>))(std::function<void ()>)>::__call<void, , 0>(std::tuple<>&&, std::_Index_tuple<0>) at functional:1145
  67: void std::_Bind<void (*(std::function<void ()>))(std::function<void ()>)>::operator()<, void>() at functional:1203
  68: callable_action_instance_t<std::_Bind<void (*(std::function<void ()>))(std::function<void ()>)> >::run_action() at callable_action.hpp:31
  69: callable_action_wrapper_t::run() at runtime_utils.cc:43
  70: coro_t::run() at coroutines.cc:205
  71: void coro_t::spawn_now_dangerously<std::_Bind<void (*(std::function<void ()>))(std::function<void ()>)> >(std::_Bind<void (*(std::function<void ()>))(std::function<void ()>)>&&) at coroutines.hpp:52
  72: repeating_timer_t::on_timer() at timing.cc:99
  73: timer_handler_t::on_oneshot() at timer.cc:62
  74: timerfd_provider_t::on_event(int) at timerfd_provider.cc:86
  75: epoll_event_queue_t::run() at epoll.cc:115
  76: linux_thread_pool_t::start_thread(void*) at thread_pool.cc:155
  77: /lib/x86_64-linux-gnu/libpthread.so.0(+0x7e9a) [0x7f823669fe9a] at 0x7f823669fe9a (/lib/x86_64-linux-gnu/libpthread.so.0)
  78: clone+0x6d at 0x7f82363cc8bd (/lib/x86_64-linux-gnu/libc.so.6)
```

**Comment by timmaxw** [comment]
In CR 2948.

**Comment by timmaxw** [comment]
Merged in 558dc41c1bda08ac061b22e0f555dca12da145c2.
